### DISCUSSION TITLE
Python Type Annotations

### DISCUSSION THREAD

**Discussion by lyxal** [Vyxal/Vyxal#196] [Rewrite] [answered]
# THIS IS TO DO WITH PYTHON TYPE ANNOTATIONS

Two Questions:

**Do we want it?**

What are our opinions on variable typing?

**Do we actually need it?**

Of course, type annotations/static typing is a good idea in theory, but is it useful for Vyxal? For example, element related functions will most always have one of the following signatures:

- `Any -> Any`
- `Any, Any -> Any`
- `Any, Any, Any -> Any`

Because they can take any value and turn it into something else. So really, typing would only be used in helper functions and compilation stuff.

Is this enough usage to justify typing?

**Comment by mathcat4** [comment]
I don't think we need it for helper functions. It's recommended in some pep i think, but as helper functions have an explanation anyways (in a readable format), it seems useless.

**Comment by ysthakur** [comment, marked as answer]
I think we only need it for helper functions that are used elsewhere in the source code itself. Having type annotations for builtins with overloads is useless because it'll just be a ginormous union type or `Any`, and most of these builtins aren't being called in the source code directly, they're called in the transpiled code. *If* a type annotation can be something other than just `Any`, it shouldn't be left out.

**Comment by ghost** [comment]


**Reply by ysthakur** [reply]
Type annotations don't influence the runtime in any way (if anything, they might slow it down very slightly because the types are actual values). I do agree that they'll be helpful for understanding code, though.

**Reply by lyxal** [reply]
First of all, welcome back! It's been way too long since your last visit to SE ;)

Anyhow, this is pretty much what we decided on doing except for element functions...the function signature for each element function will always be along the lines of

```python
def name(lhs: Union[int, sympy.Rational, str, list, LazyList], rhs: Union[int, sympy.Rational, str, list, LazyList]) -> Union[int, sympy.Rational, str, list, LazyList]:
```

Meaning that in this case, having type annotations for each element function would have the opposite effect of creating readability.

**Reply by ghost** [reply]
Thank you :) I didn't realize that Julia and Python work differently.

I've been completely brainwashed by Julia type annotations, so if this means something else for Python, I apologize
